Atualiza organização de pastas

gilvan-reis opened the pull request:
# Descrição
Essa PR visa sugerir uma proposta de alteração na organização de pastas, separando os arquivos por responsabilidade e deixando o diretorio raiz com menos arquivos.

Como mencionado no commit, o diretorio `/analytics` não foi removido nessa PR porque ele é referenciado no frontend e precisamos fazer uma alteração em duas etapas para evitar que o site fique indisponivel momentaneamente.

Fiquem a vontade para comentar sobre as mudanças nos nomes das pastas

PS: decidi por implementar algumas melhorias no codigo em uma PR separada, a fim de evitar um bloqueio devido a essas alterações

# Proximos passos
- [x] Criar PR no front para apontar para o nosso diretorio de dados analisados
- [x] Documentar o proposito de cada pasta (provavelmente em uma sessão do README)

gilvan-reis committed:
fix: Corrige erro de digitação no Makefile

gilvan-reis committed:
refactor: Move codigos para pasta exclusiva

gilvan-reis committed:
refactor: Renomeia pasta de arquivos "cru"

gilvan-reis committed:
refactor: Renomeia pasta de arquivos "limpos"

gilvan-reis committed:
refactor: Renomeia pasta de arquivos analisados

A pasta /analytics não foi removido nesse commit porque ela é
referenciada no frontend. Então é necessario fazer uma alteração
em duas etapas para não quebrar o site momentaneamente.

rbelem commented on app/code/download_data.py:
```diff
@@ -19,4 +19,4 @@
     link = one_a_tag['href']
     if "Vacinados" in link:
-        urllib.request.urlretrieve(link, 'raw_db/' + link.split("/")[-1])
+        urllib.request.urlretrieve(link, 'app/data/raw/' + link.split("/")[-1])
```
Acho que ficaria melhor mover o data/ pra raíz do código e o analytics pra dentro do data.

gilvan-reis replied:
@rbelem Poderia explicar os motivos do porque você acha que essa mudança ficaria melhor? Assim fica mais facil de entender a sua linha de raciocínio para essa mudança

rbelem replied:
@gilvan-reis Acho que dessa forma ficariam melhores separados of códigos fonte dos dados e concentraria-se os dados num só lugar.

gilvan-reis replied:
A proposta atual está assim:
- `app/analysis` (diretorio com as analises geradas e utilizadas no front)
- `app/code` (diretorio com os codigos)
- `app/data/raw` (diretorio com os dados "cru" [pdf])
- `app/data/cleaned` (diretorio com os dados "limpos" [json])

E a sua proposta seria:
- `app/code/`
- `app/code/data/raw`
- `app/code/data/cleaned`
- `app/code/data/analysis`

é isso?

rbelem replied:
@gilvan-reis seria mais ou menos assim:
- app/
- data/raw
- data/cleaned
- data/analysis

gilvan-reis replied:
Sobre ter o `data/analysis` não tenho nada contra (inclusive foi a minha primeira opção). Deixei separado porque considero que ele tem um contexto de crescimento diferente dos dado (atualmente todas as analises estão sendo exportadas em csv devido a forma como é utilizado, mas existem varias outras possibilidades para exportação dessas analises). Nesse ponto, vou esperar o comentário de mais pessoas para ter uma decisão.

Sobre ter o código no diretório `app` sou desfavorável a essa modificação quando comparado com ter um diretório especifico para isso (`app/code`), visto que ele especifica o conteúdo do diretório e evita a necessidade de uma nova reestruturação do código caso a quantidade de códigos cresça

rbelem replied:
@gilvan-reis sobre o ```app/``` acho que quando o código começar a evoluir mais, será criados módulos dentro do ```app/```

No futuro ficaria assim:
- app/
- app/main.py
- ```app/parser/__init__.py```
- app/parser/parser.py

gilvan-reis replied:
Ainda acho que a opção de ter uma pasta especifica é melhor quando comparado com a solução proposta, porque teríamos responsabilidades diferentes num mesmo nível de pasta. 

Vale reforçar que essa solução proposta é a mesma apresentada no comentário acima (mas sem a pasta `code`):
> E a sua proposta seria:
> 
> * `app/code/`
> * `app/code/data/raw`
> * `app/code/data/cleaned`
> * `app/code/data/analysis`
> 
> é isso?

Como a evolução do código é a mesma nos dois cenários, vou esperar outros comentários para ver se alguém tem outro argumento sobre essa mudança.

pvfrota replied:
Na conversa de hoje, chegamos numa idéia de que seria bom guardar todos os dados históricos já processados, para fins de transparência, o que acham de já pensarmos numa estratégia de organização de diretórios que englobe isso?

gilvan-reis replied:
Poderiamos colocar os arquivos gerados dentro de um diretorio no `analysis` contendo o nome do arquivo que gerou ele, mas o maior problema seria como o front iria acessar esses dados, visto que a solução atual é hard coded.

Tambem podemos pensar na possibilidade de separar as pastas por tipo de vacina, para permitir a adição de outros tipos no futuro

pvfrota replied:
@gilvan-reis oq vc acha de começarmos a separar nossos diretórios por dia? Assim, a estrutura passaria a ser mais ou menos assim

```bash
app
├── {...}
└── data
    ├── {...}
    ├── 2020-02-01
    │   ├── {...}
    │   ├── analysis
    │   ├── cleaned
    │   └── raw
    ├── 2020-02-02
    │   ├── {...}
    │   ├── analysis
    │   ├── cleaned
    │   └── raw
    ├── 2020-02-03
    │   ├── {...}
    │   ├── analysis
    │   ├── cleaned
    │   └── raw
    └── 2020-02-04
        ├── {...}
        ├── analysis
        ├── cleaned
        └── raw
```

Assim, o front poderia ter uma lista consolidada dessas datas (diretórios), e conforme fosse escolhida uma nova data lá, ele buscava no diretório correspondente

gilvan-reis replied:
@pvfrota é uma solução possível, mas as pastas "raw" e "cleaned" teriam apenas um arquivo cada. Se a opção do front ter essa lista de datas for aceita, podemos fazer na estrutura atual da mesma forma (apenas vai ser necessário subir uma PR no back e no front para cada alteração).

Acho que seria bom você também comentar o que acha sobre os dois pontos em aberto:
- Ter uma pasta "code" para centralizar os codigos ou deixa-los na raiz da pasta
- Colocar a pasta "analysis" dentro do "data" ou deixa-la na raiz
- EXTRA: os nomes das pastas estão bons? Fiz algumas alterações porque achei que estavam mais significativos, mas é importante todos terem essa mesma percepção.

Estou esperando essas opiniões para poder atualizar a PR e fazer o merge dela

rbelem replied:
Referência interessante pra gente se basear.
https://docs.python-guide.org/writing/structure/

agnjunio commented on Dockerfile:
```diff
@@ -9,4 +9,4 @@ RUN apt-get update && \
 RUN pip install -r requirements.txt
 
-COPY . /app
+COPY . .
```
Seria interessante adicionar `WORKDIR /vacina-manaus-backend` ou algo assim antes, pra isolar os arquivos do projeto e evitar conflito com arquivos do file system do linux

gilvan-reis replied:
No Dockerfile já tem o `WORKDIR /app` na linha 3, portanto o código já está organizado como pode ser visto no print abaixo, mas acho proveitoso a mudança do nome

![image](https://user-images.githubusercontent.com/34248754/106889449-c46d7e80-66c6-11eb-8253-65ea4dab23ae.png)

gilvan-reis committed:
chore: Move arquivo de codeowners para dentro da pasta do github

gilvan-reis commented:
Devido a não termos uma conclusão sobre os pontos em aberto, decidi que:
- O código ficará na pasta "app/code" ao invés de ficar na raiz da pasta "app"
- A pasta "analysis" vai ficar em "app/data/analysis"

Tambem movi o arquivo de codeowners para dentro da pasta ".github" para deixar a raiz do projeto mais limpa (A [documentação](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners#codeowners-file-location) permite essa mudança).

Sendo assim, a PR pode ser revisada novamente baseada nessas definições

rbelem reviewed (changes requested):
@gilvan-reis não concordo com as modificações. Se não chegamos a um consenso permanece do jeito que está.

gilvan-reis commented:
@rbelem discordo da sua opinião. Se você discorda das modificações citadas é necessário que você sugira modificações e argumente porque você considera que a opção sugerida é melhor.

Se as soluções forem similares (como é o caso) e a divergência for apenas uma preferencia pessoal, acho muito prejudicial não entregar o valor da PR que melhora significativamente a organização do repositório

rbelem commented:
> @rbelem discordo da sua opinião. Se você discorda das modificações citadas é necessário que você sugira modificações e argumente porque você considera que a opção sugerida é melhor.
> 

@gilvan-reis concordo com o que você falou. Eu deveria adicionar mais argumentos. Obrigado por me apontar pra essa direção. Vou elaborar melhor.

rbelem commented:
O backend é uma aplicação feita em python. Nesse momento cada ação está separada em seu próprio arquivo e a tendência é evoluir pra um módulo. Então acho que precisamos seguir pro modo de organização sugerida por essa documentação https://docs.python-guide.org/writing/structure/
Se compararmos outras aplicações python de grandes projetos vamos perceber um padrão diferente do proposto.
O dados e código fonte são coisas diferentes e precisam ficar em diretórios separados, dessa forma o diretório 'code' ficaria redundante. Caso a proposta seja o diretório '/app' se tornar o root do projeto, também acho ser redundante, pois a raiz do projeto já o é.
Fazendo uma comparação com o sistema de arquivo unix, os diretórios de módulos dentro do diretório 'app/' seria destinado para o diretório de módulos do python(/usr/lib/pythonx.x/) e o binário principal iria para /usr/bin. Os arquivos de dados seriam instalados em /var/lib/nome-projeto. As configurações ficariam em /etc.
Se a explicação ficou estranha me avisa, que eu vou procurar melhorar e escrever de  forma diferente.

gilvan-reis commented:
Eu, o @pvfrota e o @rbelem fizemos uma reunião síncrona para resolver essas discussões.

Seguimos com a seguinte decisão:
- Considerando `/` como sendo a raiz do repositório no github, teremos o diretorio `/src` para centralizar o codigo e teremos o diretório `/data` para centralizar os dados
- A pasta `/data` terá como subdiretório uma divisão por ano, mês e dia (dia e mês com dois dígitos, ano com quatro dígitos). Dentro da pasta de dia teremos um diretório para representar cada etapa do ciclo de processamento de dados (raw, extracted, cleaned e analyzed)
- O @pvfrota vai documentar os desejos e necessidades desse ciclo de processamento de dados

gilvan-reis commented:
A fim de desbloquear essa PR e devido ao escopo de alterar o script que gera as analises não pertencer a esta PR, estou criando a pasta `data/analyzed` que centralizará todos os dados analisados utilizando a distribuição atual, deixando que a divisão em subpastas por ano, mes e dia fique para outra PR.

pvfrota commented:
@gilvan-reis vou deixar essa PR on-hold até que a gente tenha uma PR de front-end que utilize essa estrutura

gilvan-reis commented:
@pvfrota poderia deixar mais claro os motivos de deixar essa PR on hold? Quais dados são utilizados no front-end e estão impedindo a mudança? Se são apenas os dados do analytics, não tinhamos decidido por fazer essa alteração em duas etapas?

rbelem commented on src/process_data.py:
```diff
@@ -392,5 +392,5 @@ def process_all(self):
     else:
         fileName = get_latest_filename()
-    dataProcessor = DataProcessor(fileName, 'analytics')
+    dataProcessor = DataProcessor(fileName, 'data/analysis')
```
aqui deveria ser 'data/analyzed/'?

gilvan-reis replied:
Deveria. Vou corrigir

rbelem committed:
chore: Resolvendo conflito e formatando Dockerfile

gilvan-reis commented:
A atualização da localização do front foi feita na PR https://github.com/CodeForManaus/vacina-manaus/pull/54
